proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+type family GSlotTypeOf (repA :: Type) :: Type where
+  GSlotTypeOf (K1 _ c _) = c
+  GSlotTypeOf (M1 _ _ f p) = GSlotTypeOf (f p)
+  GSlotTypeOf _ = TypeError ('Text "No match for GSlotTypeOf")
+```
+
+Usage:
+
+```haskell
+ghci> :set -XDataKinds
+ghci> :kind! LiftedCtorToSpecCtor 'Left (Either a b)
+= '(Generics.MetaCons
+      "Left" PrefixI False,
+    '[ '(MetaSel Nothing NoSourceUnpackedness NoSourceStrictness DecidedLazy,
+         a)])
+ghci> :kind! LiftedCtorToSpecCtor '(,) (a, b)
+= '(MetaCons
+      "(,)" PrefixI False,
+    ['(MetaSel Nothing NoSourceUnpackedness NoSourceStrictness DecidedLazy,
+       a),
+     '(MetaSel Nothing NoSourceUnpackedness NoSourceStrictness DecidedLazy,
+       b)])
+```
+
+### Library side: what the safe interface looks like
+
+In this proposal I've tried to limit the changes to GHC to a minimum, and make most of the work on the library side. Here's a snapshot of what can be done when the new primops are integrated into [a memory-safe library](https://github.com/tweag/linear-base/pull/450).
+
+I define a custom `IdCard` type, and then build an instance of it directly in a compact region, using destinations (variables `d_`) to fill the fields of the data structure little by little:
+
+```haskell
+data IdCard = MkIdCard {name :: String, address :: Address}
+  deriving (Eq, Generic, Show)
+data Address = MkAddress {no :: Int, street :: String, city :: String, zipCode :: Int}
+  deriving (Eq, Generic, Show)
+
+testBuildDirectlyInRegion = do
+  let actual :: IdCard
+      !actual = alloc' (\d ->
+        case fill @'MkIdCard d of
+          (dN, dA) ->
+            fillLeaf "John Doe" dN
+              `lseq` (case fill @'MkAddress dA of
+                        (dNo, dSt, dC, dZ) ->
+                          fillLeaf 42 dNo
+                            `lseq` fillLeaf "Elm Street" dSt
+                            `lseq` fillLeaf "Springwood" dC
+                            `lseq` fillLeaf 12345 dZ
+                      )
+          )
+      expected :: IdCard
+      expected = MkIdCard "John Doe" (MkAddress 42 "Elm Street" "Springwood" 12345)
+  assertEqual "John Doe ID is the same in both building schemes" expected actual
+```
+
+The syntax looks quite verbose compared to naive structure building, because the API is designed to allow efficient structure building, little-by-little, from dynamic data, instead of building the whole structure at once from static data (where it is no more efficient than a copy using existing `compactAdd#`).
+
+The API uses linear types to ensure (at compile-time) that every field of the data structure is wrote to exactly once before the structure can be read. More details on the API can be found in this [JFLA 2024 article](https://inria.hal.science/hal-04406360/document) and [this blog post](https://www.tweag.io/blog/2024-03-07-dps-haskell/).
+
+### Library side: more performant S-expression parser
```

Comment:
I think I already said that, but I don't think this should be in the example section, it should be in the effect section.